File: 2019/24/24.go
package main

import (
	"fmt"
	"io/ioutil"
	"strings"
)

type Point struct {
	x, y int
}
type Grid map[Point]string

func parseFile(fileName string) Grid {
	data, _ := ioutil.ReadFile(fileName)
	grid := make(Grid)
	for y, line := range strings.Split(string(data), "\n") {
		for x, char := range strings.Split(line, "") {
			grid[Point{x, y}] = char
		}
	}
	return grid
}

func neighbours(point Point) []Point {
	x := point.x
	y := point.y
	return []Point{
		Point{x + 1, y},
		Point{x - 1, y},
		Point{x, y + 1},
		Point{x, y - 1},
	}
}

type RecursivePoint struct {
	x, y, level int
}

func recursiveNeighbours(point Point) []RecursivePoint {
	res := []RecursivePoint{}
	for _, n := range neighbours(point) {
		if n.x == 2 && n.y == 2 {
			continue
		}
		res = append(res, RecursivePoint{n.x, n.y, 0})
	}

	if point.x == 0 {
		res = append(res, RecursivePoint{1, 2, 1})
	}
	if point.y == 0 {
		res = append(res, RecursivePoint{2, 1, 1})
	}
	if point.y == 4 {
		res = append(res, RecursivePoint{2, 3, 1})
	}
	if point.x == 4 {
		res = append(res, RecursivePoint{3, 2, 1})
	}

	if point.x == 1 && point.y == 2 {
		for y := 0; y < 5; y++ {
			res = append(res, RecursivePoint{0, y, -1})
		}
	}
	if point.x == 2 && point.y == 1 {
		for x := 0; x < 5; x++ {
			res = append(res, RecursivePoint{x, 0, -1})
		}
	}
	if point.x == 3 && point.y == 2 {
		for y := 0; y < 5; y++ {
			res = append(res, RecursivePoint{4, y, -1})
		}
	}
	if point.x == 2 && point.y == 3 {
		for x := 0; x < 5; x++ {
			res = append(res, RecursivePoint{x, 4, -1})
		}
	}

	return res
}

func newValue(value string, bugs int) string {
	if value == "." && (bugs == 1 || bugs == 2) {
		return "#"
	}
	if value == "#" && bugs == 1 {
		return "#"
	}
	return "."
}

func iterateSingle(grid Grid) Grid {
	new := make(Grid)
	for point, value := range grid {
		bugs := 0
		for _, n := range neighbours(point) {
			if val, ok := grid[n]; ok {
				if val == "#" {
					bugs++
				}
			}
		}
		new[point] = newValue(value, bugs)
	}
	return new
}

func isBug(grids map[int]Grid, level int, p RecursivePoint) bool {
	grid, ok := grids[level+p.level]
	if !ok {
		return false
	}
	point := Point{p.x, p.y}
	value, ok := grid[point]
	if !ok {
		return false
	}
	return value == "#"
}

func iterateMultiple(grids map[int]Grid) map[int]Grid {
	new := make(map[int]Grid)
	for level, grid := range grids {
		new[level] = make(Grid)
		for point, value := range grid {
			if point.x == 2 && point.y == 2 {
				new[level][point] = "?"
				continue
			}
			bugs := 0
			for _, n := range recursiveNeighbours(point) {
				if isBug(grids, level, n) {
					bugs++
				}
			}
			new[level][point] = newValue(value, bugs)
		}
	}
	return new
}

func biodiversity(grid Grid) int {
	multiplier := 1
	value := 0
	for y := 0; y < 5; y++ {
		for x := 0; x < 5; x++ {
			if grid[Point{x, y}] == "#" {
				value += multiplier
			}
			multiplier *= 2
		}
	}
	return value
}

func firstRepeating(grid Grid) int {
	seen := make(map[int]bool)
	for {
		grid = iterateSingle(grid)
		bd := biodiversity(grid)
		if _, ok := seen[bd]; ok {
			return bd
		}
		seen[bd] = true
	}
}

func emptyGrid() Grid {
	g := make(Grid)
	for x := 0; x < 5; x++ {
		for y := 0; y < 5; y++ {
			g[Point{x, y}] = "."
		}
	}
	return g
}

func countBugs(grids map[int]Grid) int {
	bugs := 0
	for _, grid := range grids {
		for _, value := range grid {
			if value == "#" {
				bugs++
			}
		}
	}
	return bugs
}

func partOne() int {
	grid := parseFile("input.map")
	return firstRepeating(grid)
}

func partTwo() int {
	grids := make(map[int]Grid)
	for level := -200; level < 201; level++ {
		grids[level] = emptyGrid()
	}
	grids[0] = parseFile("input2.map")

	for i := 0; i < 200; i++ {
		grids = iterateMultiple(grids)
	}
	return countBugs(grids)
}

func main() {
	fmt.Println("Part 1:", partOne())
	fmt.Println("Part 2:", partTwo())
}

File: 2019/24/24.py
from collections import defaultdict


def parse_file(file):
    grid = {}
    with open(file) as f:
        for y, line in enumerate(f.readlines()):
            for x, c in enumerate(line.strip()):
                grid[x, y] = c
    return grid


def draw(grid):
    for y in range(5):
        for x in range(5):
            print(grid[x, y], end='')
        print()
    print()


def neighbours(point):
    x, y = point
    ns = {
        (x + 1, y, 0),
        (x - 1, y, 0),
        (x, y + 1, 0),
        (x, y - 1, 0),
    }
    ns.discard((2, 2, 0))
    if x == 0:
        ns.add((1, 2, 1))
    if y == 0:
        ns.add((2, 1, 1))
    if y == 4:
        ns.add((2, 3, 1))
    if x == 4:
        ns.add((3, 2, 1))
    if point == (1, 2):
        for y in range(5):
            ns.add((0, y, -1))
    if point == (2, 1):
        for x in range(5):
            ns.add((x, 0, -1))
    if point == (3, 2):
        for y in range(5):
            ns.add((4, y, -1))
    if point == (2, 3):
        for x in range(5):
            ns.add((x, 4, -1))
    return ns


def iterate(grids):
    new = defaultdict(dict)
    for level, grid in grids.items():
        for point, value in grid.items():
            if point == (2, 2):
                new[level][point] = '?'
                continue
            bugs = sum(
                1 for x, y, n_level in neighbours(point)
                if grids.get(level + n_level, {}).get((x, y), '.') == '#'
            )
            if value == '.':
                if bugs in [1, 2]:
                    new[level][point] = '#'
                else:
                    new[level][point] = '.'

            if value == '#':
                if bugs == 1:
                    new[level][point] = '#'
                else:
                    new[level][point] = '.'
    return new


def count_bugs(grids):
    bugs = 0
    for grid in grids.values():
        for value in grid.values():
            if value == '#':
                bugs += 1
    return bugs


grid = parse_file('input2.map')
grids = {}
for level in range(-200, 201):
    grids[level] = {point: '.' for point in grid.keys()}
grids[0] = grid

for i in range(200):
    grids = iterate(grids)

print("Part 2:", count_bugs(grids))
